lib/widgets/message_list.dart

```diff
@@ -503,24 +523,102 @@ class MessageList extends StatefulWidget {
 
+  List<int>? _messagesRecentlyInViewport;
+
+  /// Which messages are onscreen.
+  ///
+  /// Ignores outbox messages.
+  ///
+  /// A message is considered onscreen if
+  /// - it covers the full height of the viewport
+  ///   (i.e. its top is above the viewport top
+  ///   and its bottom is below the viewport bottom) or
+  /// - its bottom is in the viewport (i.e. between the viewport top and bottom)
+  ///
+  /// This definition is helpful for mark-as-read-on-scroll, because
+  /// when this method is called during a fast scroll through many messages,
+  /// the returned range (almost*) always includes at least one message.
+  /// For the mark-as-read request, we "fill in" any messages
+  /// between that message(s) and the messages currently in view.
+  /// That can be needed when the scrolling is so fast that
+  /// some messages go by without rendering in the viewport.
+  ///
+  /// (*It can be empty in the middle of the message list, rarely,
+  /// if a tall message covers most of the viewport
+  /// but its bottom isn't visible and the message above it is offscreen,
+  /// on the far side of a date separator or recipient header.)
+  List<int> _getMessagesInViewport() {
+    final messageItemElements = <Element>[];
+    void visit(Element element) {
+      final widget = element.widget;
+      if (widget is MessageItem) {
+        if (widget.item is MessageListMessageItem) {
+          messageItemElements.add(element);
+        }
+        return;
+      }
+      element.visitChildElements(visit);
+    }
+
+    final scrollViewElement = _scrollViewKey.currentContext as Element;
+    // TODO this frequently ends up walking through a few hundred elements,
+    //   only 5-10 of which end up being MessageItem ones.
+    scrollViewElement.visitChildElements(visit);
+
+    final scrollViewRenderObject = scrollViewElement.renderObject as RenderBox;
+    final viewportHeight = scrollViewRenderObject.size.height;
+
+    final result = <int>[];
+    for (final element in messageItemElements) {
+      final renderObject = element.renderObject as RenderBox;
+      final widget = element.widget as MessageItem;
+      final item = widget.item as MessageListMessageItem; // (see `visit`)
+      final message = item.message;
+
+      final messageHeight = renderObject.size.height;
+      final messageTop = renderObject.localToGlobal(
+        Offset.zero, ancestor: scrollViewRenderObject).dy;
+      final messageBottom = renderObject.localToGlobal(
+        Offset(0, messageHeight), ancestor: scrollViewRenderObject).dy;
+
+      final doesMessageSpanViewport = messageTop < 0 && messageBottom > viewportHeight;
+      if (doesMessageSpanViewport) {
+        result.add(message.id);
+        break;
+      }
+
+      final isMessageBottomVisible = messageBottom > 0 && messageBottom < viewportHeight;
+      if (isMessageBottomVisible) {
+        result.add(message.id);
+      }
+    }
+    // TODO why isn't it sorted already?
```

Comment:
This visits the items in tree order. Within each list sliver, the tree order of the items (their order as children of the sliver) is in the sliver's growth direction. So for the top sliver, the items appear from bottom to top, the reverse of their order as Zulip messages.